src/Spice86.Core/Emulator/ProgramExecutor.cs

```diff
@@ -175,4 +180,77 @@ private void Dispose(bool disposing) {
             _disposed = true;
         }
     }
+
+    /// <summary>
+    /// Informs the DOS memory manager that the memory where the program was loaded and its stack
+    /// was created is not available to allocate.
+    /// </summary>
+    /// <remarks>
+    /// The stack is allocated right after the executable and grows down. We need to make sure that
+    /// the DOS memory manager doesn't try to allocate that memory when asked via int21h or it will
+    /// overwrite and corrupt parts of the program we're executing or the stack. <br/><br/>
+    /// Arguably DosProcessManager.SetupCpuForExe() would be a better place to do this since it sets
+    /// up the stack after the executable has been loaded.. but DosMemoryManager has an instance of
+    /// DosProcessManager, not the other way around, so there isn't a clean way to do it without
+    /// creating circular references. ProgramExecutor, on the other hand, already has an instance of
+    /// Dos with both DosMemoryManager and DosProcessManager, and it is responsible for
+    /// orchestrating loading the program, so this class is the logical place to put it without
+    /// creating odd layering problems.
+    /// </remarks>
+    /// <param name="cpuState">The CPU state with the allocated stack segment.</param>
+    /// <param name="memoryManager">The DOS memory manager to reserve the memory.</param>
+    private void ReserveSpaceForExeAndStack(State cpuState, DosMemoryManager memoryManager) {
+        // We will always have a stack allocated after loading an EXE, but not necessarily after
+        // loading a COM file. That's okay. If we don't have it, we'll skip this reservation.
+        if (cpuState.SS == 0) {
+            if (_loggerService.IsEnabled(LogEventLevel.Verbose)) {
+                _loggerService.Verbose(
+                    "SS={ss} SP={sp} - skipping program and stack memory reservation",
+                    ConvertUtils.ToHex16(cpuState.SS),
+                    ConvertUtils.ToHex16(cpuState.SP));
+            }
+            return;
+        }
+
+        // After a program has been loaded but before it has been executed, its data segment will
+        // point to the beginning of the PSP, which is the first structure in memory immediately
+        // before the MCB. That's the beginning the block that we need to reserve.
+        ushort pspSegment = cpuState.DS;
+
+        // After the program has been loaded but before it has been executed, the SS:SP will
+        // temporarily be the top of the stack that is the last memory address that we need to
+        // reserve (because the stack grows down). The PSP that we retrieved above precedes the
+        // program and will be the first memory address that we need to reserve. Knowing both makes
+        // calculating the size easy.
+        uint topOfStack = MemoryUtils.ToPhysicalAddress(cpuState.SS, cpuState.SP);
```

Comment:
cpuState.StackPhysicalAddress ?

Reply:
That's neat! I didn't know that existed. I guess I should have looked at the other members of the `State` class a little closer. Sure! I can switch to using that.